[FIX] web_editor: link / unlink buttons not appearing/disappearing

adsa-odoo opened the pull request:
**Current behaviour before commit:**

-When creating a link, the 'unlink' button is not getting 
appeared in the toolbar.

-When removing a link through 'unlink' button, the 'unlink' 
 button is not getting disappeared in toolbar.

-Before this commit, `_updateEditorUI` method is
 responsible for hiding/ showing create-link
 button.

**Desired behaviour after commit:**

-Now, When creating a link, the 'unlink' button gets
 appeared in the toolbar.

-When removing link through 'unlink' button, the
 'unlink' button gets disappeared in toolbar.

-This commit ensures that `_updateToolbar` method
 is responsible for hiding/ showing the create-link 
 and unlink button.

task-3514639



---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/140262).

cammarosano commented:
Hello @adsa-odoo ! Can you solve the conflicts in this branch?

adsa-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:
```diff
@@ -2336,7 +2336,7 @@ export class OdooEditor extends EventTarget {
             }
             const returnValue = editorCommands[method](this, ...args);
-            if (link) {
+            if (link && link.isConnected) {
```
@cammarosano Indeed now it is not required as `contenteditable = false` thing is removed from the link. I have updated the PR, thanks!

cammarosano commented:
Hello! Is there a particular reason you targeted saas-16.4?
At least for Issue 2 in the task ("User will see the unlink button in the toolbar"), I see it happens (at least) since 16.0.

cammarosano commented:
So, you basically reverted [my commit](https://github.com/odoo/odoo/commit/ceedd72b893316f5994727063a6a6faf49476408) that transfered the responsability of managing button visibility from the editor to  `Wysiwyg._updateEditorUI`.

Now we have code to manage this button visibility in two different places.
Besides, the unlink button is always shown/hidden correctly (updated on selection change), while the create link button depends on mouseup...

If you're gonna undo my stuff... undo it right! :laughing: 
If you want manage the visibility of the unlink button in the _updateToolbar method, OK. But then manage also the one for create-link.

Also take care of hiding the button group as a whole if both buttons are invisible. That's what this part of my commit did:
```js
        // Clear "d-none" for button groups.
        for (const buttonGroup of this.toolbarEl.querySelectorAll('.btn-group')) {
            buttonGroup.classList.remove('d-none');
        }
        ...
                // Hide button groups that have no visible buttons.
        for (const buttonGroup of this.toolbarEl.querySelectorAll('.btn-group:not(.d-none)')) {
            if (!buttonGroup.querySelector('.btn:not(.d-none)')) {
                buttonGroup.classList.add('d-none');
            }
        }

```
While it's generic, it was written with the link button group in mind. You might able to remove this from `Wysiwyg._updateEditorUI` without breaking other stuff (please test thoroughly!) and handle the link button group case in `_updateToolbar`.

And you might need to handle the unlink button visilbility when clicked on media case.

Give it a try and see if it works. But don't stress too much about. It might not be possible after all without a refactor of this toolbar buttons managing... (which we are not going to do on stable)

cammarosano commented:
>And you might need to handle the unlink button visilbility when clicked on media case.

Something like this could be useful, for you to know in `_updateToolbar` whether the toolbar is in "media mode" or not:
![image](https://github.com/odoo/odoo/assets/65516503/daa30698-5e31-4a6b-9aef-78edf7602594)
It's from a [PR](https://github.com/odoo/odoo/pull/136687) on master...

adsa-odoo commented:
@cammarosano Hello, regarding target version, on 16.0 I am able to reproduce issue 2) "Unlink button still visible" but not the first one. Morover, I added two lines in `_updateToolbar` because when link is created or removed  `_updateEditorUI` is not getting called :smile:. Since it is causing bug I changed it but do you think that we need to revert your whole commit and transfer other piece of code into `_updateToolbar`?

adsa-odoo commented:
@cammarosano I just checked in 16.0 that we can't get selection on link after creation of the link. But when I tried this  `Promise.resolve().then(() => setSelection(link, 0, link, link.childNodes.length, false))` in callback function after save, it worked.

cammarosano commented:
Alright, let's keep this for 16.4+ then.
 
> do you think that we need to revert your whole commit and transfer other piece of code into _updateToolbar?

Yes, pretty much.
Do not do a literal git revert, but in `_updateToolbar` you should handle the visibility of the both link buttons (and the button group), and none of that should be done in `Wysiwyg._updateEditorUI`.
In `Wysiwyg._updateEditorUI` you should toggle an "oe_media" class in the toolbar when isInMedia, and use that information in `_updateToolbar` to decide whether the unlink button should be visible.

Make you sure you test Project, Knowledge, Website and mass_mailing after you make changes to the toolbar. 

And also please write a unit test so we avoid future regressions.
`addons/web_editor/static/tests/html_field_tests.js` is a good place for it, ping me if you need help with that. You should be able to make assertions based on the visibility of the buttons in the toolbar.
Check also `addons/web_editor/static/tests/field_html_tests.js` on 16.0 (they were accidentaly removed, they are not there in 16.4...) for inspiration for those tests!

adsa-odoo commented:
Alright! I think I got it, thanks!

adsa-odoo commented:
@cammarosano  Hello, I've almost done the changes you mentioned in previous comment but I have a doubt that why do we need to hide unlink button on media? If the selection is not in the link then unlink button will still not be visible, no matter whether it is media or something else. Am I missing something?

cammarosano commented:
![image](https://github.com/odoo/odoo/assets/65516503/c08b5964-a661-434a-b947-aac79c856419)
That's a very good question.
As I see it, the selection could be the result of a click on a image and still be inside a link (we can have IMG elements inside A elements).
When we click on an image, we have a totally different toolbar, dedicated to image edition: shape, crop, size, etc... And most of the text-editing buttons are not present on this "media toolbar".
One could argue that we would like to unlink an image... but we're not gonna change the current spec on stable. So, no create-link and unlink buttons when the toolbar is on "inMedia" mode! :smile:

adsa-odoo commented:
@cammarosano Hello, I have done the changes, Can you have a look?

cammarosano commented on addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:
```diff
@@ -3094,13 +3094,26 @@ export class OdooEditor extends EventTarget {
         linkButton && linkButton.classList.toggle('active', !!linkNode);
+        // hide unlink button if no link in selection, always hide on media elements
+        const unlinkButton = this.toolbar.querySelector('#unlink');
+        unlinkButton && unlinkButton.classList.toggle('d-none', isInMedia || !linkNode);
```
You could do `unlinkButton?.classList.toggle('d-none', isInMedia || !linkNode)`, just like you did for the the `linkButton` and the `linkButtonGoup` in the lines below. (this is optional, as you wish)

cammarosano commented on addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:
```diff
@@ -3094,13 +3094,26 @@ export class OdooEditor extends EventTarget {
         redoButton && redoButton.classList.toggle('disabled', !this.historyCanRedo());
+
+        // hide create-link button if selection spans several blocks, always hide on media elements
+        const range = getDeepRange(this.editable, { sel, correctTripleClick: true });
```
Why do you need to use `getDeepRange` here?

adsa-odoo replied:
Good question!, Let's consider this scenario, `<p>[test]</p><p><123</test>`. If user did selection through triple click, then `spansBlocks` will be true but selection is still there at the first paragraph. So I though correcting triple click will be better here. What's your opinion here? Should we remain it as it was before?

cammarosano replied:
Ok! I think you're right! Let's keep it like this.

cammarosano commented on addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:
```diff
@@ -3094,13 +3094,26 @@ export class OdooEditor extends EventTarget {
         redoButton && redoButton.classList.toggle('disabled', !this.historyCanRedo());
+
+        // hide create-link button if selection spans several blocks, always hide on media elements
+        const range = getDeepRange(this.editable, { sel, correctTripleClick: true });
+        const spansBlocks = descendants(range.commonAncestorContainer).some(node => isBlock(node));
```
Kudos for using `.some` instead of that`.filter ... length` thing here! Nice!! :+1: 
I think you could use simply `.some(isBlock)` here if you want.

adsa-odoo replied:
Ah! Actually I forgot that using `childNodes` will do the same thing :smiley:. Nevermind, I will change it from `descendants` to `childNodes` .

cammarosano commented:
> @cammarosano Hello, I have done the changes, Can you have a look?

Nice! Good job! I have a few remarks (questions about details I might not understand, actually!), but we're almost done. The result (behavior) is very good! :+1:

adsa-odoo commented:
> > @cammarosano Hello, I have done the changes, Can you have a look?
> 
> Nice! Good job! I have a few remarks (questions about details I might not understand, actually!), but we're almost done. The result (behavior) is very good! 👍

Thank you!! I have done the requred changes!

cammarosano commented:
@Goaman I reviewed this PR, and for me it's ready for r+!

Goaman commented:
robodoo r+
